orderer/common/follower/follower_chain.go

```diff
@@ -426,6 +426,129 @@ func (c *Chain) pullUpToJoin() error {
+	if err != nil {
+		return errors.Wrap(err, "failed to determine channel membership from last config")
+	}
+
+	for !channelMember {
+		configBlock, err := c.pullUntilNextConfig()
+		if err != nil {
+			return errors.Wrap(err, "failed to pull until next config")
+		}
+
+		channelMember, err = c.clusterConsenter.IsChannelMember(configBlock)
+		if err != nil {
+			return errors.Wrap(err, "failed to determine channel membership from pulled config")
+		}
+
+		c.logger.Debugf("next config block: %d, channel member: %v", c.lastConfig.Header.Number, channelMember)
+	}
+
+	return err
+}
+
+// pullUntilNextConfig will return the next config block or an error.
+func (c *Chain) pullUntilNextConfig() (*common.Block, error) {
+	// This puller is built from the tip of the ledger. When we get a new config block we exit the method. That will
+	// destroy the puller. Next time we enter, a new puller is built, taking in the new config.
+	puller, err := c.createBlockPullerFunc()
+	if err != nil {
+		c.logger.Errorf("Error creating block puller: %s", err)
+		return nil, errors.Wrap(err, "error creating block puller")
+	}
+	defer puller.Close()
+
+	var configBlock *common.Block
+	for configBlock == nil {
+		endpoint, latestHeight, err := cluster.LatestHeightAndEndpoint(puller)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get latest height and endpoint")
+		}
+		c.logger.Debugf("Orderer endpoint %s has the biggest ledger height: %d", endpoint, latestHeight)
+
+		if latestHeight <= c.ledgerResources.Height() {
+			c.logger.Debugf("My height: %d, latest height: %d; going to wait %v for latest height to grow",
+				c.ledgerResources.Height(), latestHeight, c.retryInterval)
+			select {
+			case <-c.stopChan:
+				return nil, errors.New("stopped while waiting for latest height to grow")
+			case <-c.timeAfter(c.retryInterval):
+				c.increaseRetryInterval()
+				continue
+			}
+		}
+
+		configBlock, err = c.pullUntilLatestOrConfig(puller, latestHeight)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to pull until latest height or config")
+		}
+		c.resetRetryInterval()
+	}
+
+	return configBlock, nil
+}
+
+func (c *Chain) pullUntilLatestOrConfig(puller ChannelPuller, latestHeight uint64) (*common.Block, error) {
+	firstBlockToPull := c.ledgerResources.Height()
+	var nextBlock = c.ledgerResources.Block(firstBlockToPull - 1)
+	if nextBlock == nil {
+		return nil, errors.Errorf("cannot retrieve previous block %d", firstBlockToPull-1)
+	}
+	var actualPrevHash = protoutil.BlockHeaderHash(nextBlock.Header)
+
+	// Pull until the latest height or a config block
+	for seq := firstBlockToPull; seq < latestHeight; seq++ {
+		select {
+		case <-c.stopChan:
+			c.logger.Warnf("Stopped while pulling blocks: from %d until %d, last pulled %d",
+				firstBlockToPull, latestHeight, nextBlock.Header.Number)
+			return nil, errors.New("stopped while pulling blocks")
+		default:
+			nextBlock = puller.PullBlock(seq)
```

Comment:
Out of curiosity, how is this implemented? I hope that it's not a separate API call per block and leverages an underlying stream instead?

Reply:
Internally, the block puller sends a seek info of:
```
        orderer.SeekInfo{
		Start:         &orderer.SeekPosition{Type: &orderer.SeekPosition_Specified{Specified: &orderer.SeekSpecified{Number: startSeq}}},
		Stop:          &orderer.SeekPosition{Type: &orderer.SeekPosition_Specified{Specified: &orderer.SeekSpecified{Number: math.MaxUint64}}},
		Behavior:      orderer.SeekInfo_BLOCK_UNTIL_READY,
		ErrorResponse: orderer.SeekInfo_BEST_EFFORT,
	}
```
It then reads blocks from the stream, until its internal in-memory buffer (20MB by default) is full, and then returns the first block read. Subsequent calls will just pop blocks from the buffer until it is empty.

Reply:
On the first BlockPull(n), the BlockPuller will connect to some orderer that has at least block n, and will inquire what is its latest block as well. It will then try to fill a buffer with with as many blocks between [n,latest], up to a maximal buffer size. On the next BlockPull(n+1), it will pop from the buffer, until the buffer is empty. When the buffer is empty, it all starts again. for pulling a batch of blocks, it uses a stream.

Reply:
Addressed. Never dropping blocks.